core/src/main/java/hudson/security/GlobalSecurityConfigurationInclude.java

```diff
@@ -0,0 +1,129 @@
+            configureGlobalSecurityDefaults(wiz);
+            globalSecurityConfiguration = new SetupWizardSecurityConfiguration(wiz);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+    
+    private void configureGlobalSecurityDefaults(Jenkins j) {
+        HudsonPrivateSecurityRealm securityRealm = new HudsonPrivateSecurityRealm(false, false, null);
+        //securityRealm.setForceFirstUserCreation(false); // this will take over
+        j.setSecurityRealm(securityRealm);
+        j.setDisableRememberMe(true);
+        FullControlOnceLoggedInAuthorizationStrategy authStrategy = new FullControlOnceLoggedInAuthorizationStrategy();
+        j.setAuthorizationStrategy(authStrategy);
+        try {
+            j.setSlaveAgentPort(-1); // -1 to disable
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        
+        // this can be set globally
+        Jenkins.getInstance().setCrumbIssuer(new DefaultCrumbIssuer(false));
+    }
+    
+    public GlobalSecurityConfiguration getGlobalSecurityConfiguration() {
+        return globalSecurityConfiguration;
+    }
+    
+    public synchronized void doConfigure(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, FormException {
+        // This may configure with SetupWizard
+        SecurityContext oldContext = ACL.impersonate(ACL.SYSTEM);
+        try {
+            globalSecurityConfiguration.configure(req, req.getSubmittedForm());
+            
+            // move config to global jenkins
+            if(Jenkins.getInstance() instanceof SetupWizard) {
+                SetupWizard setup = (SetupWizard)Jenkins.getInstance();
+                setup.complete(); // This sets up the *real* jenkins app
+                // and then move the security settings over and save
+                BulkChange bc = new BulkChange(Jenkins.getInstance());
+                try{
+                    ((SetupWizard)globalSecurityConfiguration.getJenkins()).applySecuritySettings(Jenkins.getInstance());
+                    Jenkins.getInstance().save(); // now, actually save the configuration
+                    if(Jenkins.getInstance().getSecurityRealm() instanceof HudsonPrivateSecurityRealm
+                            && !(Jenkins.getInstance().getAuthorizationStrategy() instanceof Unsecured)) {
+                        // send to create a user
+                        FormApply.success(req.getContextPath()+"/securityRealm/firstUserDirect").generateResponse(req, rsp, null);
+                    }
+                    else {
+                        // redirect to jenkins home
+                        FormApply.success(req.getContextPath()+"/").generateResponse(req, rsp, null);
+                    }
+                } finally {
+                    bc.commit();
+                }
+                globalSecurityConfiguration = new GlobalSecurityConfiguration();
+            }
```

Comment:
You got the Jenkins instance at the start of this block and assigned it to `setup`, then you are calling other functions (more than one and multiple times) to get the the Jenkins instance. Is that really needed, or couldn't you just use `setup` everywhere? If it is needed, then that seems to smell like a problem.